### PR TITLE
perl: exclude version patch number in lib directory

### PR DESCRIPTION
**Pull request by cho-m**
- [x] Have you followed the [guidelines for contributing](https://github.com/Homebrew/homebrew-core/blob/HEAD/CONTRIBUTING.md)?
- [x] Have you ensured that your commits follow the [commit style guide](https://docs.brew.sh/Formula-Cookbook#commit)?
- [ ] Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?
- [ ] Have you built your formula locally with `brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [ ] Is your test running fine `brew test <formula>`, where `<formula>` is the name of the formula you're submitting?
- [ ] Does your build pass `brew audit --strict <formula>` (after doing `brew install --build-from-source <formula>`)? If this is a new formula, does it pass `brew audit --new <formula>`?

-----

Splitting off the change from #84191.

Directory structure using only major/minor version to avoid breakage on Perl patch releases. This is closer to how Arch Linux and macOS package it:
- **Arch** https://archlinux.org/packages/core/x86_64/perl/
   - usr/lib/perl5/5.34/core_perl/CORE/libperl.so
- **macOS**
   - /Library/Developer/CommandLineTools/SDKs/MacOSX11.3.sdk/System/Library/Perl/5.30/darwin-thread-multi-2level/CORE/libperl.tbd 

---

One thing that may be worth considering is setting up an installation structure for Perl formulae similar to other Linux distros and Homebrew's Python formulae so that we can easily pick up very common packages like XML::Parser rather than jumping through hoops with the `PERL5LIB`.

### CODE REVIEW

**Review by SMillerDev — DISMISSED** [review]

**Comment by Bo98** [comment]
I've not looked into the grokj2k situation yet, but based on this PR alone this makes sense to me.

This is a backwards incompatible change, though would've eventually broke anyway when updating to 5.34.1 (if there even will be one).

> One thing that may be worth considering is setting up an installation structure for Perl formulae similar to other Linux distros and Homebrew's Python formulae so that we can easily pick up very common packages like XML::Parser rather than jumping through hoops with the `PERL5LIB`.

I'm open to some restructuring if it makes things easier but I'm not sure how that solves the issue described. `XML::Parser` does not ship with the Perl formula. We ship it on Linux with `intltool` and it requires `PERL5LIB` because it is installed into libexec.

**Comment by cho-m** [comment]
The latter part is partly comparing to how Linux distros may package Perl modules and installed them into general `vendor_perl`. Also, how we have a few commonly used Python packages as formulae that install into general prefix rather than venv/libexec.

However, based on Homebrew's rules relating to CPAN modules, it probably doesn't align with formulae requirements. Overall, not too important of a topic and can be ignored.

I'm just noting that there is often some overhead when interacting with Perl formulae.

---

On separate note, after Linux CI finishes, I need to check on the `libperl` linkage for the revision-bumped formulae.

For example, current `vim` bottle on Linux only has following `perl` paths in RPATH
- `/home/linuxbrew/.linuxbrew/Cellar/perl/5.34.0/lib/perl5/5.34.0/x86_64-linux-thread-multi/CORE`
- `/home/linuxbrew/.linuxbrew/opt/perl/lib`

`libperl.so` is linked via the full versioned Cellar. Since I didn't change `vim` formula other than revision bump, I think this will still be an issue. If so, I need to see if the path can be manually set to `opt` path rather than the Cellar path being autodetected.

Also want to check `weechat` and `mhonarc` (and any other formulae with broken linkage that may be output from Linux CI).

**Comment by Bo98** [comment]
In theory, that could potentially be fixed by doing one of:

* Installing `libperl` directly into lib.
  - Perl's build system does not support doing this.
* Try `-Dprefix` to opt and `-Dinstallprefix` to Cellar.
  - `-Dprivlib` etc that we customise would also be set to the opt. Not sure if `-Dinstallprivlib` is necessary - it might just work automatically based on `-Dprefix` and `-Dinstallprefix`.
* Hack `Config.pm` and/or `Config_heavy.pl` if all options fail.

**Comment by cho-m** [comment]
I think Debian does a combination of first and last https://salsa.debian.org/perl-team/interpreter/perl/-/blob/debian-5.32/debian/rules#L168. Their libraries are provided as:
- Package: **libperl5.32**
   - /usr/lib/x86_64-linux-gnu/libperl.so.5.32
   - /usr/lib/x86_64-linux-gnu/libperl.so.5.32.1
- Package: **libperl-dev**
   - /usr/lib/x86_64-linux-gnu/libperl.a
   - /usr/lib/x86_64-linux-gnu/libperl.so

---

I'll probably try the 2nd option for now as it seems simpler and a smaller change from current bottles.

Though, it retains the demerit of some annoying libperl path, which requires a bit more work to get to once when we need it in a dependent formula (e.g. this is what I did in `grokj2k` to update RPATH).

**Comment by cho-m** [comment]
I locally tried adding `installprefix` for Linuxbrew and paths are now:
```console
# perl -V:'install(privlib|archlib|vendorlib|vendorarch|sitelib|sitearch)'
installarchlib='/home/linuxbrew/.linuxbrew/Cellar/perl/5.34.0/lib/perl5/5.34/x86_64-linux-thread-multi';
installprivlib='/home/linuxbrew/.linuxbrew/Cellar/perl/5.34.0/lib/perl5/5.34';
installsitearch='/home/linuxbrew/.linuxbrew/Cellar/perl/5.34.0/lib/perl5/site_perl/5.34/x86_64-linux-thread-multi';
installsitelib='/home/linuxbrew/.linuxbrew/Cellar/perl/5.34.0/lib/perl5/site_perl/5.34';
installvendorarch='';
installvendorlib='';

# perl -V:'(privlib|archlib|vendorlib|vendorarch|sitelib|sitearch)'
archlib='/home/linuxbrew/.linuxbrew/opt/perl/lib/perl5/5.34/x86_64-linux-thread-multi';
privlib='/home/linuxbrew/.linuxbrew/opt/perl/lib/perl5/5.34';
sitearch='/home/linuxbrew/.linuxbrew/opt/perl/lib/perl5/site_perl/5.34/x86_64-linux-thread-multi';
sitelib='/home/linuxbrew/.linuxbrew/opt/perl/lib/perl5/site_perl/5.34';
vendorarch='';
vendorlib='';
```

---

In comparison, the macOS system Perl:
```console
❯ /usr/bin/perl -V:'install(privlib|archlib|vendorlib|vendorarch|sitelib|sitearch)'
installarchlib='/Library/Perl/Updates/5.30.2/darwin-thread-multi-2level';
installprivlib='/Library/Perl/Updates/5.30.2';
installsitearch='/Library/Perl/5.30/darwin-thread-multi-2level';
installsitelib='/Library/Perl/5.30';
installvendorarch='/Network/Library/Perl/5.30/darwin-thread-multi-2level';
installvendorlib='/Network/Library/Perl/5.30';

❯ /usr/bin/perl -V:'(privlib|archlib|vendorlib|vendorarch|sitelib|sitearch)'
archlib='/System/Library/Perl/5.30/darwin-thread-multi-2level';
privlib='/System/Library/Perl/5.30';
sitearch='/Library/Perl/5.30/darwin-thread-multi-2level';
sitelib='/Library/Perl/5.30';
vendorarch='/Network/Library/Perl/5.30/darwin-thread-multi-2level';
vendorlib='/Network/Library/Perl/5.30';
```

File: Formula/perl-build.rb [hunk 1]
```diff
@@ -125,6 +126,15 @@ def install
...
       (bin/cmd).write_env_script(libexec/"bin/#{cmd}", PERL5LIB: ENV["PERL5LIB"])
     end
+
+    # Some scripts' shebang is set to the realpath of Homebrew perl, which is not
+    # found using `detected_perl_shebang`. Also, since Formula["perl"].bin returns
+    # the opt_bin directory, we need to manually construct the Cellar path.
+    on_linux do
+      inreplace Dir[libexec/"bin/{perl-build,config_data}"] do |s|
+        s.sub! %r{^#!#{HOMEBREW_PREFIX}/Cellar/perl/[^/]+/bin/perl}, "#!#{Formula["perl"].opt_bin}/perl"
+      end
```
Comment (cho-m) [inline Formula/perl-build.rb:136]: I think the `perl-build` build script replaces the Perl shebang with realpath of Homebrew Perl. I couldn't find another way to deal with this:
- `detected_perl_shebang` will only return system `/usr/` Perl paths
- Trying to inreplace `Formula["perl"].bin` fails as it returns opt_bin:
    ```
    Error: An exception occurred within a child process:
      Utils::Inreplace::Error: inreplace failed
    /home/linuxbrew/.linuxbrew/Cellar/perl-build/1.32/libexec/bin/perl-build:
      expected replacement of #<Pathname:/home/linuxbrew/.linuxbrew/opt/perl/bin> with #<Pathname:/home/linuxbrew/.linuxbrew/opt/perl/bin>
    /home/linuxbrew/.linuxbrew/Cellar/perl-build/1.32/libexec/bin/config_data:
      expected replacement of #<Pathname:/home/linuxbrew/.linuxbrew/opt/perl/bin> with #<Pathname:/home/linuxbrew/.linuxbrew/opt/perl/bin>
    ```

File: Formula/mhonarc.rb [hunk 2]
```diff
@@ -24,9 +24,6 @@ class Mhonarc < Formula
   end
 
   def install
-    # Handle the hardcoded binary script
-    inreplace "mhonarc", "#!/usr/bin/perl", "#!/usr/bin/env perl"
```
Comment (cho-m) [inline Formula/mhonarc.rb:28]: Removed this as it should be automatically inreplaced with Homebrew perl shebang.

File: Formula/perl-build.rb [hunk 3]
```diff
@@ -125,6 +126,15 @@ def install
...
       (bin/cmd).write_env_script(libexec/"bin/#{cmd}", PERL5LIB: ENV["PERL5LIB"])
     end
+
+    # Some scripts' shebang is set to the realpath of Homebrew perl, which is not
+    # found using `detected_perl_shebang`. Also, since Formula["perl"].bin returns
+    # the opt_bin directory, we need to manually construct the Cellar path.
+    on_linux do
+      inreplace Dir[libexec/"bin/{perl-build,config_data}"] do |s|
+        s.sub! %r{^#!#{HOMEBREW_PREFIX}/Cellar/perl/[^/]+/bin/perl}o, "#!#{Formula["perl"].opt_bin}/perl"
+      end
```
Comment (cho-m) [inline Formula/perl-build.rb:136]: **EDIT:** Re-added comment after force-push.

I think the `perl-build` build script replaces the Perl shebang with realpath of Homebrew Perl. I couldn't find another way to deal with this:
- `detected_perl_shebang` will only return system `/usr/` Perl paths
- Trying to inreplace `Formula["perl"].bin` fails as it returns opt_bin:
    ```
    Error: An exception occurred within a child process:
      Utils::Inreplace::Error: inreplace failed
    /home/linuxbrew/.linuxbrew/Cellar/perl-build/1.32/libexec/bin/perl-build:
      expected replacement of #<Pathname:/home/linuxbrew/.linuxbrew/opt/perl/bin> with #<Pathname:/home/linuxbrew/.linuxbrew/opt/perl/bin>
    /home/linuxbrew/.linuxbrew/Cellar/perl-build/1.32/libexec/bin/config_data:
      expected replacement of #<Pathname:/home/linuxbrew/.linuxbrew/opt/perl/bin> with #<Pathname:/home/linuxbrew/.linuxbrew/opt/perl/bin>
    ```

**Comment by cho-m** [comment]
#### freeradius-server

Due to `collectd` being installed on CI runner before `freeradius-server`, it is being detected by configure. I can't find a way to prevent it and have tried all of: `--without-collectd`, `--without-collectdclient`, `--with-collectd=no`, `--with-collectdclient=no`, `--with-collectdclient-lib-dir=`, `--with-collectdclient-lib-dir=no`, `--with-collectdclient-lib-dir=/dev/null`

Since most repos don't enable this feature and Debian specifically has comment on there being issues (https://salsa.debian.org/debian/freeradius/-/blob/master/debian/control#L6), I want to avoid this being linked in new bottle. At this point, I may just need to manually inreplace the `configure` or `configure.ac`.

#### mhonarc

With Perl directory changes, `mhonarc` is now failing to pick the correct directories. It is now trying to set directories relative to `prefix` (e.g. `prefix/man` rather than `prefix/share/man` and `prefix/lib` rather than `prefix/lib/perl5/site_perl/VERSION/. Not sure exact reason. The related logic may be:
```perl
    my $cfg_prefix        = interpolate_path($Config{'prefix'});
    $DefValues{'binpath'} = interpolate_path($Config{'installbin'});
    $DefValues{'libpath'} = interpolate_path($Config{'installsitelib'});
    $DefValues{'manpath'} = interpolate_path($Config{'installman1dir'});
    if (defined($OptValues{'prefix'})) {
	$DefValues{'binpath'} = join($DIRSEP, $OptValues{'prefix'}, 'bin')
	    unless $DefValues{'binpath'} =~
		   s/^\Q$cfg_prefix/$OptValues{'prefix'}/o;
	$DefValues{'libpath'} = join($DIRSEP, $OptValues{'prefix'}, 'lib')
	    unless $DefValues{'libpath'} =~
		   s/^\Q$cfg_prefix/$OptValues{'prefix'}/o;
	$DefValues{'manpath'} = join($DIRSEP, $OptValues{'prefix'}, 'man')
	    unless $DefValues{'manpath'} =~
		   s/^\Q$cfg_prefix/$OptValues{'prefix'}/o;
	$DefValues{'docpath'} = join($DIRSEP, $OptValues{'prefix'}, 'doc');
```

**Comment by cho-m** [comment]
Linux failures (`Error: 52 failed steps!`) are mostly CI issues and previously seen issues. May need to make sure about `rbenv`, but everything else seems unrelated.
```
brew test --retry --verbose rbenv
```

<details>
<summary>21 due to reoccurring Java CI issue</summary>

```
brew test --retry --verbose aspectj
brew test --retry --verbose bbtools
brew test --retry --verbose beagle
brew test --retry --verbose bnfc
brew test --retry --verbose cfr-decompiler
brew test --retry --verbose dita-ot
brew test --retry --verbose duck
brew test --retry --verbose gradle
brew test --retry --verbose groovy
brew test --retry --verbose groovysdk
brew test --retry --verbose hadoop
brew test --retry --verbose htmlcleaner
brew test --retry --verbose javacc
brew test --retry --verbose jdnssec-tools
brew test --retry --verbose jmeter
brew test --retry --verbose joshua
brew test --retry --verbose jruby
brew test --retry --verbose kaitai-struct-compiler
brew test --retry --verbose leiningen
brew test --retry --verbose tika
brew test --retry --verbose umple
```
</details>

<details>
<summary>24 due to no space left on CI runner</summary>

```
brew install --only-dependencies grpc-swift
brew install grpc-swift
brew install infer
brew install --only-dependencies lima
brew install lima
brew install llvm@7
brew install llvm@8
brew install llvm@9
brew install mercury
brew install mingw-w64
brew test --retry --verbose nexus
brew install nushell
brew test --retry --verbose pyoxidizer
brew install qemu
brew install root
brew install swift
brew install --only-dependencies swift-protobuf
brew install swift-protobuf
brew install --only-dependencies swift-sh
brew install swift-sh
brew install tesseract-lang
brew install --only-dependencies --include-test tree-sitter
brew test --retry --verbose tree-sitter
brew install widelands
```
</details>

<details>
<summary>6 failures also seen in Python 3.9.7 PR #84283</summary>

```
brew test --retry --verbose buku
brew test --retry --verbose julia
brew test --retry --verbose kafka
brew test --retry --verbose mrbayes
brew test --retry --verbose parquet-cli
brew test --retry --verbose streamlink
```
</details>

File: Formula/perl.rb [hunk 4]
```diff
@@ -30,26 +31,24 @@ class Perl < Formula
...
       -Dstartperl=#!#{opt_bin}/perl
-      -Dman1dir=#{man1}
-      -Dman3dir=#{man3}
+      -Dman1dir=#{opt_share}/man/man1
+      -Dman3dir=#{opt_share}/man/man3
       -Duseshrplib
       -Duselargefiles
       -Dusethreads
     ]
     args << "-Dsed=/usr/bin/sed" if OS.mac?
```
Comment (carlocab) [inline Formula/perl.rb:47]: We can remove this, as it's a no-op. Could probably wait for a follow-up PR though.

**Review by carlocab — APPROVED** [review]

File: Formula/perl.rb (hunk 4 above)
Comment (cho-m) [reply Formula/perl.rb:47]: I'll add this with I get a chance to rebase. `vim` was updated so the PR needs to be re-run for new bottles.

**Pull request closed without merging**